File: Sources/SomnioProtocol/Payloads/Session.swift
import Foundation

/// Redeem a stored session token in place of a password login. Accepted **pre-login only**:
/// it is an alternative to `LoginMessage`, not something an attached connection can replay.
///
/// The outcome reuses `LoginResultMessage` rather than introducing a redeem-specific response.
/// That keeps the new-tag count down and, more importantly, makes an expired, unknown, and
/// revoked token indistinguishable to the client — all three answer `.badCredentials`, so a
/// probing client learns nothing about which tokens ever existed.
public struct RedeemSessionMessage: Codable, Sendable, Equatable {
    public var token: String

    public init(token: String) {
        self.token = token
    }
}

/// Revoke the token presented on this connection. Accepted **post-attach only**.
///
/// Revocation deliberately does not ride ordinary disconnect cleanup: refresh, network loss,
/// and a normal socket close must all *preserve* the token, since resumption across exactly
/// those events is the point of the feature. Only an explicit logout destroys it. Scope is the
/// presented token alone — "log out everywhere" is a different feature.
public struct RevokeSessionMessage: Codable, Sendable, Equatable {
    public var token: String

    public init(token: String) {
        self.token = token
    }
}

/// Sent only in response to a `Login` that asked for one. A successful `RedeemSession` resolves the
/// presented token and mints no replacement — redemption deliberately does not rotate, because
/// rotating before `WorldRouter.register` succeeds would invalidate the credential on the first
/// `alreadyLoggedIn` answer and break the bounded resume retry. So no frame of this kind follows a
/// resume, and a client must not wait for one. The raw token is returned exactly once; the server
/// keeps only a digest.
public struct SessionTokenMessage: Codable, Sendable, Equatable {
    public var token: String
    /// Seconds until expiry. `Int32` because the 30-day lifetime is 2,592,000 seconds, well
    /// past `Int16`. Sending a duration rather than an absolute instant means the client never
    /// has to trust its own clock offset.
    public var expiresInSeconds: Int32

    public init(token: String, expiresInSeconds: Int32) {
        self.token = token
        self.expiresInSeconds = expiresInSeconds
    }
}

/// Acknowledgement that the presented token's row is gone. Request-gated like every other
/// session frame: the server sends it only in response to a `RevokeSessionMessage`.
public struct SessionRevokedMessage: Codable, Sendable, Equatable {
    public var revoked: Bool

    public init(revoked: Bool) {
        self.revoked = revoked
    }
}

File: packages/server/src/handlers/session.ts
import { LOGIN_RESULT, SOMNIO_PROTOCOL_CONSTANTS, utf8ByteLength } from '@somnio/protocol'
import type { RedeemSessionMessage, RevokeSessionMessage } from '@somnio/protocol'
import type { ConnectionActor } from '../connection/connectionActor.ts'
import type { ConnectionDependencies } from '../connection/dependencies.ts'
import type { ConnectionOutbox } from '../connection/outbox.ts'
import type { Logger } from '../logging.ts'
import { completeAuthenticatedJoin, sendLoginResult } from './login.ts'

/**
 * Redeeming a token in place of a password login. Every failure — unknown, expired, revoked —
 * answers `badCredentials`, deliberately telling the client nothing about which. An over-cap
 * token is refused before the repository is reached, so an unauthenticated frame cannot drive an
 * unbounded digest.
 */
export async function handleRedeem(
  message: RedeemSessionMessage,
  connection: ConnectionActor,
  dependencies: ConnectionDependencies
): Promise<void> {
  const outbox = connection.outbox
  const logger = dependencies.logger
  if (utf8ByteLength(message.token) > SOMNIO_PROTOCOL_CONSTANTS.maxSessionTokenUTF8Bytes) {
    sendLoginResult(outbox, LOGIN_RESULT.badCredentials, logger)
    return
  }
  try {
    const resolved = await dependencies.sessions.redeem(message.token)
    if (resolved === undefined) {
      sendLoginResult(outbox, LOGIN_RESULT.badCredentials, logger)
      return
    }
    // No rotation on redemption: rotating before `worldRouter.register` succeeds would invalidate the
    // credential on the first `alreadyLoggedIn` answer and break the client's bounded retry.
    await completeAuthenticatedJoin(resolved.accountId, connection, dependencies, false)
  } catch (error) {
    logger.error({ error: String(error) }, 'session redemption failed')
    sendLoginResult(outbox, LOGIN_RESULT.badCredentials, logger)
  }
}

/**
 * Revocation is its own request-gated message with an acknowledgement, never a side effect of
 * disconnect. The connection's own account scopes the delete, so a player cannot revoke another
 * account's session by presenting a token they merely obtained.
 */
export async function handleRevoke(
  message: RevokeSessionMessage,
  accountId: string,
  connection: ConnectionActor,
  dependencies: ConnectionDependencies
): Promise<void> {
  const outbox = connection.outbox
  const logger = dependencies.logger
  if (utf8ByteLength(message.token) > SOMNIO_PROTOCOL_CONSTANTS.maxSessionTokenUTF8Bytes) {
    sendSessionRevoked(outbox, false, logger)
    return
  }
  try {
    sendSessionRevoked(outbox, await dependencies.sessions.revoke(message.token, accountId), logger)
  } catch (error) {
    logger.error({ error: String(error) }, 'session revocation failed')
    sendSessionRevoked(outbox, false, logger)
  }
}

function sendSessionRevoked(outbox: ConnectionOutbox, revoked: boolean, logger: Logger): void {
  outbox.sendEncoded({ tag: 'sessionRevoked', payload: { revoked } }, logger)
}
